docsrc/imap/developer/major-releasing.rst

```diff
@@ -0,0 +1,402 @@
+=====================================
+
+.. contents::
+
+These instructions describe the process of turning what was the master
+branch into a new release series, and making the first release from it.
+
+For normal point releases, see :ref:`imap-developer-releasing`
+
+Prerequisites
+=============
+
+Same as for normal releases.
+
+Feature freeze
+==============
+
+This is the period where new features are not being merged to the master
+branch.  It usually starts at the start of December, and continues until
+the new series has its own cyrus-imapd-x.y branch, which is usually forked
+in early January.
+
+Once the new series has its own branch, any bug fixes will need to be
+developed against master and then backported to the new branch.  In
+contrast, bug fixes that happen during the feature freeze only need to land
+once, so take advantage of this window to focus on those.
+
+The feature freeze also reduces races between the release manager doing the
+various tasks of a release and other developers merging changes.
+
+Once the new series has its own branch, normal feature development can resume
+on master -- developing features for the *next* major release.
+
+Make sure master is good
+========================
+
+With the master branch checked out and up to date:
+
+1. Ensure your git repository is clean, using something like
+   ``git clean -xfd``.  Note that this command will destroy any uncommitted
+   work you might have, so make sure your ducks are in line before proceeding.
+2. Generate a configure script: ``autoreconf -i -s``
+3. Generate everything else: ``./configure --enable-maintainer-mode`` (you do
+   not need any other options at this stage).
+4. Run ``make distcheck``.  This will generate a distribution tarball, and
+   test it in various ways.  It takes about 10-15 mins to run, depending on
+   your hardware.  If you usually build Cyrus with a script that sets PATH etc,
+   you will need to provide the same environment at this step.  For example,
+   ellie uses an alias like this for this step:
+
+   ``alias distcheck="PATH=/usr/local/cyruslibs/bin:$PATH make distcheck"``.
+
+   If ``make distcheck`` fails, you are not ready to proceed -- fix the
+   problems, get them tested and committed, then restart this testing.
+5. ``make distcheck`` can only test so much (it doesn't know about cunit or
+   cassandane), so you also need to check the tarball against those.
+
+   i.    The tarball will be called something like
+         ``cyrus-imapd-3.0.0-rc2-23-g0241b22.tar.gz``
+         (this corresponds to the ``git describe`` output).
+   ii.   Extract it: ``tar xfz cyrus-imapd-*.tar.gz``
+         (substitute version for ``*``).
+   iii.  Change into the directory: ``cd cyrus-imapd-*``
+   iv.   Configure it: ``./configure [...]`` (provide the same arguments and
+         environment that you would when building for Cassandane at any other
+         time).
+   v.    Compile it: ``make -j4`` -- it should build correctly.
```

Comment:
Pleade remove -j4.  This is a very personal option depnding on the used hardware amd other circumstances.

Reply:
The `-j4` argument in the examples illustrates that we _do_ support parallel builds (since without it, one might mistakenly assume we don't).  It is my hope that anyone following these instructions knows how to use `make`, knows their system, and can choose a suitable `-j` value (or none) for themselves.

We don't use `make dist` anymore, nowadays we use `make distcheck`.  One of the many things it checks is that everything that needs to be included has been included.